client/rest/src/main/java/org/elasticsearch/client/Cancellable.java

```diff
@@ -0,0 +1,72 @@
+/*
```

Comment:
I think that the apache HttpClient already provides cancellation. 
I sketched the solution here (I did not take into account a request could span multiple nodes, so the business logic has to be improved).
The basic is that the `execute` in the async client

https://github.com/apache/httpasyncclient/blob/4.1.x/httpasyncclient/src/main/java/org/apache/http/nio/client/HttpAsyncClient.java#L68

returns an implementation of `Future`  called `org.apache.http.concurrent.BasicFuture` that is also 
`org.apache.http.concurrent.Cancellable` and delegate the `cancel` to the `cancel(boolean)` already present in `Future`.

Also calling the `cancel` on this future notifies the listener, and you could exploit it to bake some custom business logic.

I did not try my solution end to end though.

Here is the branch if you want to have a look at this approach.
https://github.com/jesinity/elasticsearch/tree/rest-async-cancellation-token